hydra-main/src/main/java/com/addthis/hydra/query/aggregate/QueryTask.java

```diff
@@ -136,6 +150,29 @@ private void rescheduleSelfWithBackoff() {
         }
     }
 
+    private FrameReadTaskSourceProvider createTaskSourceProvider(MeshSourceAggregator sourceAggregator) {
+        Query query = sourceAggregator.query;
+        int totalTasks = sourceAggregator.totalTasks;
+        int maxSimul = getMaxSimul(query.getParameter("maxSimul"), totalTasks);
+        if (maxSimul == totalTasks) {
+            return new DefaultFrameReadTaskSourceProvider(sourceAggregator);
+        } else {
+            return new MaxSimulFrameReadTaskSourceProvider(sourceAggregator, maxSimul);
+        }
+    }
+
+    private int getMaxSimul(@Nullable String maxSimulQueryParam, int totalTasks) {
+        if (maxSimulQueryParam == null) {
+            return totalTasks;
+        }
+        try {
```

Comment:
I think this can go either way - ignoring invalid maxSimul will not cause any problem, other than the fact that task sources are not activated eagerly. I basically opted for less code.